tests/js/server/aql/aql-optimizer-conditions-merging.js

```diff
@@ -0,0 +1,195 @@
+        assertEqual(0, res.length, buildContext());
+      } else {
+        assertNotEqual(0, res.length);
+        let v = res[0];
+        if (minInclusive) {
+          assertTrue(v >= min, { v, min }, buildContext());
+        } else {
+          assertTrue(v > min, { v, min }, buildContext());
+        }
+        v = res[res.length - 1];
+        if (maxInclusive) {
+          assertTrue(v <= max, { v, max }, buildContext());
+        } else {
+          assertTrue(v < max, { v, max }, buildContext());
+        }
+      }
+
+      let nodes = AQL_EXPLAIN(query).plan.nodes.filter((n) => n.type === 'IndexNode');
+      assertEqual(1, nodes.length);
+      let condition = nodes[0].condition;
+      if (emptyCondition) {
+        assertEqual({}, condition, buildContext());
+      } else {
+        assertEqual('n-ary or', condition.type, condition, buildContext());
+        let subNodes = condition.subNodes;
+        assertEqual(1, subNodes.length);
+        assertEqual('n-ary and', subNodes[0].type, condition, buildContext());
+        subNodes = subNodes[0].subNodes;
+        let expectedConditions = 0;
+        if (minSet) {
+          let cond = subNodes[expectedConditions];
+          if (min === max) {
+            assertEqual('compare ==', cond.type);
+          } else if (minInclusive) {
+            assertEqual('compare >=', cond.type);
+          } else {
+            assertEqual('compare >', cond.type);
+          }
+          assertEqual(2, cond.subNodes.length, cond);
+          assertEqual('attribute access', cond.subNodes[0].type);
+          assertEqual('value', cond.subNodes[1].type);
+          assertEqual(min, cond.subNodes[1].value);
+          ++expectedConditions;
+        } 
+        if (maxSet && (min !== max)) {
+          let cond = subNodes[expectedConditions];
+          if (maxInclusive) {
+            assertEqual('compare <=', cond.type);
+          } else {
+            assertEqual('compare <', cond.type);
+          }
+          assertEqual(2, cond.subNodes.length, cond);
+          assertEqual('attribute access', cond.subNodes[0].type);
+          assertEqual('value', cond.subNodes[1].type);
+          assertEqual(max, cond.subNodes[1].value);
+          ++expectedConditions;
+        }
+        assertEqual(expectedConditions, subNodes.length, buildContext());
+      }
+    }
+  };
+
+  return {
+    setUpAll : function () {
+      let c = db._create(cn);
+      let docs = [];
+      for (let i = minValue; i <= maxValue; ++i) {
+        docs.push({ value: i });
+      }
+      c.insert(docs);
+      c.ensureIndex({ type: "persistent", fields: ["value"] });
+    },
+
+    tearDownAll : function () {
+      db._drop(cn);
+    },
+
+    testConditionsMerging2Conditions : function () {
+      runTests(100, 2);
+    },
+    
+    testConditionsMerging3Conditions : function () {
+      runTests(100, 3);
+    },
+    
+    testConditionsMerging4Conditions : function () {
+      runTests(100, 4);
+    },
```

Comment:
Nice idea to do some random testing!

One question on this: Do you know (e.g. by testing) how many (percent) of the generated conditions are empty, especially for 4 conditions? I could imagine it's relatively high (haven't done a calculation though), and if so, we might want to tweak the random generation so it gets less likely (and if not, its fine).

Reply:
Definitely, the more condition parts we have, the more of the conditions yield empty results.
I did a few runs and counted the percentage of empty conditions. On average we get:
* 25% empty conditions for 2 condition parts
* 55% empty conditions for 3 condition parts
* 66% empty conditions for 4 condition parts
That's far from perfect, but I think it's still a good enough test for now.
Adjusting the condition generation so that less empty conditions are produced is doable, but also not 100% trivial, so I would prefer to not do it right now.

Reply:
Agreed, the test is already pretty thorough this way. Thanks for investigating!